Issue #594 - Update documentation for `get_pairwise_comparisons()`

nikosbosse opened the pull request:
<!-- Thanks for opening this pull request! Below we have provided a suggested template for PRs to this repository and a checklist to complete before opening a PR -->
 
## Description

This PR closes #594.
This PR closes #506.
This PR closes #517.

#594 mentions a few planned updates to the documentation of `add_pairwise_comparisons()`. 

This PR
- applies the desired changes to `get_pairwise_comparisons()` instead. `add_relative_skill()` (previously `add_pairwise_comparison()` links to `get_pairwise_comparisons()`. 
- adds a figure to the docs and restructures them to make it clear what happens
- there was one particular piece of information which I dropped: namely that relative skill scores would be different depending on whether or not a baseline was present. I actually changed that quite a while ago in this commit: https://github.com/epiforecasts/scoringutils/commit/f5c73832aa38df9ccc0874fc322abd71f474f8c0. 

One proposal that we talked about, but haven't made a final decision on, was splitting the p-value part out of the function. On the one hand, computing p-values also needs the pairwise comparison mechanism. On the other hand, p-values are very different from mean score ratios and relative skill scores. 
I think splitting the p-values and the rest would require some effort. We could just drop the p-values now and re-introduce them later? 

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title as follows: *issue-number*: PR title
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] I have built the package locally and run rebuilt docs using roxygen2.
- [x] My code follows the established coding standards and I have run `lintr::lint_package()` to check for style issues introduced by my changes. 
- [ ] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

<!-- Thanks again for this PR - @scoringutils dev team -->

sbfnk commented on R/pairwise-comparisons.R:
```diff
@@ -1,24 +1,53 @@
-#' computed by [score()]. Adding unrelated columns will change results in an
-#' unpredictable way.
+#'
+#' The input should be a `scores` object as produced by [score()]. Note that
+#' adding additional unrelated columns can unpredictably change results.
```
Is there any particular reason we're worried about that here and/or is there a way that we can check it ourselves rather than telling the user?

nikosbosse replied:
I think we can't check that - this would require storing the forecast unit at the time it was computed and comparing it with the forecast unit that we see now. 
There is nothing in the suggested workflows that would lead to this situation. But I can imagine users doing some computations on their scores and doing pairwise comparisons afterwards - that's why I thought the note would be helpful.

sbfnk replied:
In that case, can we just add why it can unpredictably change results?

nikosbosse replied:
Updated to the following: 

"Note that adding additional unrelated columns can unpredictably change results, as all present columns are taken into account when determining the set of overlapping forecasts between two models"

sbfnk commented on R/pairwise-comparisons.R:
```diff
@@ -1,24 +1,53 @@
+#' The input should be a `scores` object as produced by [score()]. Note that
+#' adding additional unrelated columns can unpredictably change results.
+#' The output of the pairwise comparisons is a set of mean score ratios,
+#' relative skill scores and p-values.
```
maybe also add a description to @return? Something like
```r
#' @return A data.table with the results of pairwise comparisons containing the
#' mean score ratios (`mean_scores_ratio`), unadjusted (`pval`) and adjusted
#' (`adj_pval`) p-values, and relative skill values of each model
#' (`..._relative_skill`). If a baseline model is given then the scaled
#' relative skill is also reported as (`..._scaled_relative_skill`).
```

nikosbosse replied:
Done

sbfnk commented on R/pairwise-comparisons.R:
```diff
@@ -1,24 +1,53 @@
+#' In addition, the function computes p-values for the comparison between two
+#' models (again based on the set of overlapping forecasts). P-values can be
+#' computed in two ways: based on a nonparametric Wilcoxon signed-rank test
+#' (internally using [wilcox.test()] with `paired = TRUE`) or based on a
+#' permutation test. The permutation test is based on the difference in mean
+#' scores between two models. The null hypothesis is that the mean score
+#' difference is zero (see [permutation_test()]).
+#' Adjusted p-values are computed by calling [p.adjust()] on the raw p-values.
```
```suggestion
#' (internally using [stats::wilcox.test()] with `paired = TRUE`) with a null hypothesis that the ratio of scores of the two models is symmetric around 1 or, alternatively, based on a
#' permutation test of the difference in mean
#' scores between two models, with a null hypothesis is that the mean score
#' difference is zero (see [permutation_test()]).
#' The method can be selected using the `test_type` argument, which is passed to [compare_two_models()].
#' 
#' Adjusted p-values are computed by calling [stats::p.adjust()] on the raw p-values.
```

"null hypothesis is that the mean score difference is zero"
Can we just replace that by mean ratio being 1? If yes then that would link better with the score we're actually calculating. 

Also given the prominence of the test type in the documentation perhaps make `test_type` an explicit argument rather than passing it via ellipsis?

nikosbosse replied:
`permutation_test()` has an option to choose either difference or mean ratio == 1 🙈 The default is mean difference (although I have to admit I don't have a good intuition for how much they would differ in practice or whether they should differ at all. If I remember correctly, there was something related to whether you have a one-sided or a two-sided p-value and in some cases both p-values coincide. I remember some conversations with Johannes about this.

nikosbosse replied:
Would you be ok with leaving this for a separate issue?

nikosbosse replied:
Made a new issue, #769.

sbfnk reviewed (dismissed):
Much improved!

nikosbosse commented:
@sbfnk do you have thoughts on the broader question of whether the p-value computation should be there at all? 

> One proposal that we talked about, but haven't made a final decision on, was splitting the p-value part out of the function. On the one hand, computing p-values also needs the pairwise comparison mechanism. On the other hand, p-values are very different from mean score ratios and relative skill scores.
I think splitting the p-values and the rest would require some effort. We could just drop the p-values now and re-introduce them later?

sbfnk commented:
I think as long as the name of the function suggests that its main purpose is to do a pairwise comparison of models (and not: to calculate mean score ratios) it makes sense to keep them in as one of several aspects of pairwise comparison.

nikosbosse commented:
I think there is a case to be made for reducing code complexity splitting up the two (and maybe then renaming the function to `get_score_ratios()` or something like that. At the moment `plot_pairwise_comaprisons()` also has two separate modes, one for mean score ratios and one for p-values. It just feels a bit inelegant. But I see your point as well.

sbfnk commented:
Should this question be discussed in a separate Issue?

nikosbosse commented:
> Should this question be discussed in a separate Issue?

Yes makes sense. I was just thinking that in case we all thought it would be best to remove the p-value computation for now then that would spare me from addressing some of your comments above. But I guess it makes sense to address them now and then reiterate.

sbfnk commented:
You could keep the PR open (perhaps convert to draft) whilst we discuss the other question?

nikosbosse commented:
@seabbs Seb is on holiday. I addressed his concerns by 
- clarifying why having a superfluous column is problematic when computing pairwise comparisons (it messes up the set of overlapping forecasts)
- making the return statement more verbose (implementing Seb's suggestion)
- opening a new issue, #769, to address his points that maybe some arguments should not be buried in `...`

Are you happy for me to merge this PR?
